mkosi/machine.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-License-Identifier: LGPL-2.1+
+
+from __future__ import annotations
+
+import contextlib
+import enum
+import signal
+import sys
+from subprocess import PIPE, CompletedProcess
+from typing import Any, List, Optional, Sequence, Union
+
+import pexpect  # type: ignore
+
+from . import (
+    MKOSI_COMMANDS_SUDO,
+    build_stuff,
+    bump_image_version,
+    check_native,
+    check_output,
+    check_root,
+    init_namespace,
+    load_args,
+    needs_build,
+    parse_args,
+    prepend_to_environ_path,
+    print_output_size,
+    run_command_image,
+    run_qemu_cmdline,
+    run_shell_cmdline,
+    save_manifest,
+    suppress_stacktrace,
+    unlink_output,
+    virt_name,
+)
+from .backend import _FILE, die, run
+
+
+class Machine:
+    def __init__(self, args: Optional[Sequence[str]] = None) -> None:
+        # Remains None until image is built and booted, then receives pexpect process
+        self._serial: Optional[pexpect.spawn] = None
+        self.stack = contextlib.ExitStack()
+        self.args = load_args(parse_args(args)["default"])
+        self.exit_code: int = 1
+
+
+    @property
+    def serial(self) -> pexpect.spawnu:
+        if self._serial is None:
+            raise ValueError("Trying to access serial console before machine boot or after machine shutdown.")
+
+        return self._serial
+
+
+    def __enter__(self) -> Machine:
+        if self.args.verb in MKOSI_COMMANDS_SUDO:
+            check_root()
+            unlink_output(self.args)
+
+        if self.args.verb == "build":
+            check_output(self.args)
+
+        if needs_build(self.args):
+            check_root()
+            check_native(self.args)
+            init_namespace(self.args)
+            manifest = build_stuff(self.args)
+
+        with contextlib.ExitStack() as stack:
+            prepend_to_environ_path(self.args.extra_search_paths)
+
+            if self.args.verb in ("shell", "boot"):
+                cmdline = run_shell_cmdline(self.args)
+            elif self.args.verb == "qemu":
+                # We must keep the temporary file opened at run_qemu_cmdline accessible, hence the context stack.
+                cmdline = stack.enter_context(run_qemu_cmdline(self.args))
+            else:
+                die("No valid verb was entered.")
+
+            cmdline = [str(x) for x in cmdline]
+            cmdline = " ".join(cmdline)
+
+            self._serial = pexpect.spawnu(cmdline, logfile=None, timeout=240)
+            self._serial.expect("#")
+            self.stack = stack.pop_all()
+
+        return self
+
+
+    def run(self, commands: List[str], stdout: _FILE = sys.stdout, stderr: _FILE = sys.stderr) -> CompletedProcess:
```

Comment:
No need for `stdout` and `stderr` parameters here. We're always going to want to use `PIPE` here.